File: iosApp/iosApp/Modules/Degen ✅/DegenWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3lib

// MARK: - DegenWireframeFactory

protocol DegenWireframeFactory {
    func make(_ parent: TabBarController) -> DegenWireframe
}

// MARK: - DefaultDegenWireframeFactory

final class DefaultDegenWireframeFactory {
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let cultProposalsWireframeFactory: CultProposalsWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let degenService: DegenService
    private let networksService: NetworksService

    init(
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        cultProposalsWireframeFactory: CultProposalsWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        degenService: DegenService,
        networksService: NetworksService
    ) {
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.cultProposalsWireframeFactory = cultProposalsWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.degenService = degenService
        self.networksService = networksService
    }
}

extension DefaultDegenWireframeFactory: DegenWireframeFactory {

    func make(_ parent: TabBarController) -> DegenWireframe {
        DefaultDegenWireframe(
            parent,
            currencySwapWireframeFactory: currencySwapWireframeFactory,
            cultProposalsWireframeFactory: cultProposalsWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            degenService: degenService,
            networksService: networksService
        )
    }
}

// MARK: - Assembler

final class DegenWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> DegenWireframeFactory in
            DefaultDegenWireframeFactory(
                currencySwapWireframeFactory: resolver.resolve(),
                cultProposalsWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                degenService: resolver.resolve(),
                networksService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/degen/DegenWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.degen

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.cultproposals.CultProposalsWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.degen.DegenWireframe
import com.sonsofcrypto.web3walletcore.services.degen.DegenService
import smartadapter.internal.extension.name

interface DegenWireframeFactory {
    fun make(parent: Fragment?): DegenWireframe
}

class DefaultDegenWireframeFactory(
    private val degenService: DegenService,
    private val networksService: NetworksService,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val cultProposalsWireframeFactory: CultProposalsWireframeFactory,
): DegenWireframeFactory {

    override fun make(parent: Fragment?): DegenWireframe =
        DefaultDegenWireframe(
            parent?.let { WeakRef(parent) },
            degenService,
            networksService,
            currencySwapWireframeFactory,
            cultProposalsWireframeFactory,
        )
}

class DegenWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(DegenWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultDegenWireframeFactory(
                it.resolve(DegenService::class.name),
                it.resolve(NetworksService::class.name),
                it.resolve(CurrencySwapWireframeFactory::class.name),
                it.resolve(CultProposalsWireframeFactory::class.name),
            )
        }
    }
}